src/common/wayfinder.ts

```diff
@@ -0,0 +1,255 @@
+
+    const targetGateway =
+      filteredGateways[randomInt(0, filteredGateways.length)];
+    if (targetGateway === undefined) {
+      throw new Error('No target gateway found');
+    }
+    return new URL(
+      `${targetGateway.settings.protocol}://${targetGateway.settings.fqdn}:${targetGateway.settings.port}`,
+    );
+  }
+}
+
+export class PriorityGatewayStrategy implements WayfinderRoutingStrategy {
+  public readonly name = 'priority';
+  private ario: AoARIORead;
+  private limit: number;
+  private sortBy: 'totalDelegatedStake' | 'startTimestamp' | 'operatorStake';
+  private sortOrder: 'asc' | 'desc';
+  private blocklist: string[];
+  constructor({
+    ario,
+    limit = 1,
+    sortBy = 'operatorStake',
+    sortOrder = 'desc',
+    blocklist = [],
+  }: {
+    ario: AoARIORead;
+    limit: number;
+    sortBy: 'totalDelegatedStake' | 'operatorStake' | 'startTimestamp';
+    sortOrder: 'asc' | 'desc';
+    blocklist: string[];
+  }) {
+    this.ario = ario;
+    this.limit = limit;
+    this.sortBy = sortBy;
+    this.sortOrder = sortOrder;
+    this.blocklist = blocklist;
+  }
+
+  // TODO: builder pattern to easily change the parameters for the strategy
+
+  async getTargetGateway(): Promise<URL> {
+    const { items: gateways } = await this.ario.getGateways({
+      sortOrder: this.sortOrder,
+      sortBy: this.sortBy,
+      limit: 100, // filter it after get the results as the contract does not support filters
+    });
+
+    // filter out gateways that are not joined
+    const filteredGateways = gateways
+      .filter((gateway) => gateway.status === 'joined')
+      .filter((gateway) => !this.blocklist.includes(gateway.settings.fqdn))
+      .slice(0, this.limit - 1);
+
+    const targetGateway =
+      filteredGateways[randomInt(0, filteredGateways.length)];
+
+    if (targetGateway === undefined) {
+      throw new Error('No target gateway found');
+    }
+
+    return new URL(
+      `${targetGateway.settings.protocol}://${targetGateway.settings.fqdn}:${targetGateway.settings.port}`,
+    );
+  }
+}
+
+export type WayfinderRoutingStrategyName = 'random' | 'priority' | 'fixed';
+
+export const WayfinderRoutingStrategies: Record<
+  WayfinderRoutingStrategyName,
+  new (...args: any[]) => WayfinderRoutingStrategy
+> = {
+  random: RandomGatewayStrategy,
+  priority: PriorityGatewayStrategy,
+  fixed: FixedGatewayStrategy,
+} as const;
+
+export const arnsRegex = /^[a-z0-9_-]{1,51}$/;
+export const txIdRegex = /^[a-z0-9]{43}$/;
+
+// TODO: introduce wayfinder http wrapper class/interface so we can support a variety of http clients to proxy requests through
+export type AnyFunction = (...args: any[]) => any;
+export type WayfinderHttpClient<T extends AnyFunction> = T;
+export const createWayfinderHttpClient = <T extends AnyFunction>({
+  httpClient,
+}: {
+  httpClient: T; // generic http client parameters
+}): WayfinderHttpClient<T> => {
+  return new Proxy(httpClient, {
+    async apply(
```

Comment:
we could potentially just provide the dependencies to `Wayfinder` here and expose those function to get a wayfinder HTTP client, which allows you to provide any supported `ar://` route directly to it, and exposes `wayfinder.getRedirectUrl` on the fetch client

Reply:
moved to `createWayfinderClient` which accepts a `resolveUrl` function that is used to get the redirect URL. this gives us flexibility to introduce new clients (e.g. rpc) and allows us to add data verification